upgrade/v1beta1/upgrade.go

```diff
@@ -0,0 +1,210 @@
+	}
+	return toConditions, nil
+}
+
+func CanUpgrade(config *restclient.Config, namespace string) (bool, error) {
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return false, errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	setting, err := lhClient.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameDefaultEngineImage), metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("setting %v not found", string(types.SettingNameDefaultEngineImage))
+			return true, nil
+		}
+
+		return false, errors.Wrap(err, fmt.Sprintf("unable to get setting %v", string(types.SettingNameDefaultEngineImage)))
+	}
+
+	return false, errors.Wrapf(err, "unable to upgrade from %v directly", setting.Value)
+}
+
+func upgradeVolumes(namespace string, lhClientV1beta1 *lhclientset.Clientset, lhClient *lhclientset.Clientset) error {
+	volumes, err := lhClientV1beta1.LonghornV1beta1().Volumes(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to list volumes")
+	}
+	for _, old := range volumes.Items {
+		new := &longhorn.Volume{}
+
+		if err := copier.Copy(&new, &old); err != nil {
+			return err
+		}
+
+		conditions, err := copyConditions(old.Status.Conditions)
+		if err != nil {
+			return errors.Wrap(err, "upgrade: copy volume conditions")
+		}
+		new.Status.Conditions = conditions
+
+		new, err = lhClient.LonghornV1beta2().Volumes(namespace).UpdateStatus(context.TODO(), new, metav1.UpdateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "failed to update for %v status %v", old.Kind, old.Name)
+		}
+	}
+	return nil
+}
+
+func upgradeEngineImages(namespace string, lhClientV1beta1 *lhclientset.Clientset, lhClient *lhclientset.Clientset) error {
+	engineImages, err := lhClientV1beta1.LonghornV1beta1().EngineImages(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to list engine images")
+	}
+	for _, old := range engineImages.Items {
+		new := &longhorn.EngineImage{}
+
+		if err := copier.Copy(&new, &old); err != nil {
+			return err
+		}
+
+		conditions, err := copyConditions(old.Status.Conditions)
+		if err != nil {
+			return errors.Wrap(err, "upgrade: copy engineImage conditions")
+		}
+		new.Status.Conditions = conditions
+
+		new, err = lhClient.LonghornV1beta2().EngineImages(namespace).UpdateStatus(context.TODO(), new, metav1.UpdateOptions{})
+		if err != nil {
+			if !apierrors.IsConflict(err) {
+				return errors.Wrapf(err, "failed to update for %v status %v", old.Kind, old.Name)
+			}
+			logrus.Warnf("%v: update status for %v %v v1beta2 result in conflict, skipping", upgradeLogPrefix, old.Kind, old.Name)
+			continue
+		}
```

Comment:
The conflicts caused by other kinds of CRs should be returned as well. Similar below.